### PR TITLE
Defer create CompilationTracker._buildLock

### PR DESCRIPTION
**Pull request by ToddGrun**
Locally, I see nearly all created CompilationTrackers not end up using this field and it is showing up in a speedometer allocation profile. Going to run speedometer on this to see if it helps the test.

*** from speedometer profile before change:
image

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 1]
```diff
@@ -395,6 +395,12 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
                         if (state is FinalCompilationTrackerState finalState)
                             return finalState;
 
+                        if (_buildLock is null)
+                        {
+                            var buildLock = new SemaphoreSlim(initialCount: 1);
+                            Interlocked.CompareExchange(ref _buildLock, buildLock, null);
+                        }
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:402]:
```suggestion
                        var buildLock = InterlockedOperations.Initialize(
                            ref _buildLock,
                            static () => new SemaphoreSlim(initialCount: 1));
```

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 2]
```diff
@@ -40,7 +40,7 @@ private partial class CompilationTracker : ICompilationTracker
             private CompilationTrackerState? _stateDoNotAccessDirectly;
 
             // guarantees only one thread is building at a time
-            private readonly SemaphoreSlim _buildLock = new(initialCount: 1);
+            private SemaphoreSlim? _buildLock = null;
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:43]:
```suggestion
            private SemaphoreSlim? _buildLock;
```

**Review by sharwell — APPROVED** [review]

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
i'm fine with this.  does perfview indicate an improvment?

**Comment by ToddGrun** [comment]
> i'm fine with this. does perfview indicate an improvment?

Speedometer tests are in progress, but take forever.

**Comment by ToddGrun** [comment]
Escalating out of draft status due to positive results from speedometer run.

Test Insertion PR [here](https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/530900)

Speedometer results [here](https://devdiv.visualstudio.com/DevDiv/_apps/hub/ms-vseng.pit-vsengPerf.pit-hub?targetBuild=34622.186.dn-bot.240223.034003.530900&targetBranch=34622.186.dn-bot.240223.034003.530900&targetPerfBuildId=9123640&runGroup=Speedometer&since=2024-02-23&baselineBuild=34622.186&baselineBranch=main)

Test: CompletionTest.Completion
Scenario: 9990.Totals
Counter: CLR_BytesAllocated_NonDevenv

Scenario Analysis Tab -> TypesAllocated_NonDevenv dropdown

servicehub.roslyncodeanalysisservice_System.Threading.SemaphoreSlim
  baseline: 16,561,336 bytes
  target: 0 bytes

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**